### PR TITLE
Cleanup git-log filter option handling

### PR DESCRIPTION
**Pull request by gerhardol**
Follow up to #10076 and #9555
Regression in master

## Proposed changes

Only branches were included with message filter and first stash.
This meant that other refs than branches were not shown in the grid, especially notes/commits. Other exclusions are Gerrit refs like refs/for/ and refs/changes/
A side effect is also that git notes were not shown together with the commits that has notes.

For message filtering this was no big deal, but that deactivating "Show latest stash" also deactivated "Git notes" is a bigger issue.

Note: notes/commits refs should probably never be displayed in the grid (other than with reflog), but that requires more changes.

The filter option handling was very hard to follow, this has been cleaned up.

## Screenshots 

No filter, notes and latest stash shown

image

Latest stash deactivated below

### Before

image

### After

image

## Test methodology 

Tests are updated

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/RevisionReader.cs [hunk 1]
```diff
@@ -19,18 +19,23 @@ namespace GitCommands
...
+
+        // Exclude some refs from --all
+        NoStash = 0x008, // --exclude=refs/stash
+        NoGitNotes = 0x010, // --not --glob=notes --not
+
+        NoMerges = 0x020, // --no-merges
+        FirstParent = 0x040, // --first-parent
+        SimplifyByDecoration = 0x080, // --simplify-by-decoration
+        Boundary = 0x100, // --boundary
+        Reflogs = 0x200, // --reflog
```
Comment (RussKie) [inline GitCommands/RevisionReader.cs:38]: Please keep indentation

**Review by RussKie — COMMENTED** [review]
LGTM

File: GitCommands/RevisionReader.cs (hunk 1 above)
Comment (gerhardol) [reply GitCommands/RevisionReader.cs:38]: Msvs keeps changing at every edit so it is quite annoying to retain spaces...

Comment (gerhardol) [reply GitCommands/RevisionReader.cs:38]: Can we agree on this?
This is a (minor) regression for 4.0, but also one dependency for stash-in-grid that I will suggest to be included in 4.0 too...

Comment (gerhardol) [reply GitCommands/RevisionReader.cs:38]: Rebased without changes

File: GitCommands/RevisionReader.cs [hunk 2]
```diff
@@ -19,18 +19,23 @@ namespace GitCommands
...
-        Tags                    = 0x004,    // --tags
-        Stashes                 = 0x008,    //
-        All                     = 0x00F,    // --all
-        Boundary                = 0x010,    // --boundary
-        ShowGitNotes            = 0x020,    // --not --glob=notes --not
-        NoMerges                = 0x040,    // --no-merges
-        FirstParent             = 0x080,    // --first-parent
-        SimplifyByDecoration    = 0x100,    // --simplify-by-decoration
-        Reflogs                 = 0x200,    // --reflog
+        None = 0x000,
```
Comment (mstv) [inline GitCommands/RevisionReader.cs:22]: After removing the whitespace (-0), the `#pragma warning dis/enable SA1025` should be removed, too.

File: GitCommands/RevisionReader.cs [hunk 3]
```diff
@@ -19,18 +19,23 @@ namespace GitCommands
...
+        Branches = 0x001, // --branches=<filter>
+
+        // Unused
+        // Remotes = 0x002, // --remotes
+        // Tags = 0x004, // --tags
+        All = 0x007, // --all (default is HEAD)
+
+        // Exclude some refs from --all
+        NoStash = 0x008, // --exclude=refs/stash
+        NoGitNotes = 0x010, // --not --glob=notes --not
```
Comment (mstv) [inline GitCommands/RevisionReader.cs:32]: Which name `NoGitNotes` (new) or `ShowGitNotes` (old) matches the arguments `--not --glob=notes --not`? Just the name was changed, neither the comment nor the usage. Looks like the old name was inverse / wrong.

Comment (gerhardol) [reply GitCommands/RevisionReader.cs:32]: > Looks like the old name was inverse / wrong.

Yes

**Comment by gerhardol** [comment]
Planned to make a slightly cleanup, but got hit by test bugs. I have a clue what is causing the issues, plan to create a separate PR

```text
Cleanup git-log filter option handling

* Only branches were included with message filter and first stash

* Clarify refFilter options related to branch filters.

* Exclude boundary commits also for DiffContent searches.

* Improve the ToolTip with current filter with other settings like
"first parents" and reflog.

* Handle "FilteredBranches" with empty selection consistently as "all"
branches (was occasionally the same as CurrentBranchOnly).
```

File: GitCommands/RevisionReader.cs [hunk 4]
```diff
@@ -15,24 +15,27 @@
 
 namespace GitCommands
 {
-#pragma warning disable SA1025 // Code should not contain multiple whitespace in a row
     [Flags]
     public enum RefFilterOptions
```
Comment (RussKie) [inline GitCommands/RevisionReader.cs:19]: Please move this to its own file and retain the indentation

File: IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormBrowseTests.cs [hunk 5]
```diff
@@ -153,9 +153,9 @@ public void Filters_should_behave_as_expected()
                         form.GetTestAccessor().ToolStripFilters.SetBranchFilter("Branch2");
                         WaitForRevisionsToBeLoaded(form);
                         // Assert
-                        AppSettings.BranchFilterEnabled.Should().BeTrue();
+                        //// TBD timing issue AppSettings.BranchFilterEnabled.Should().BeTrue();
                         AppSettings.ShowCurrentBranchOnly.Should().BeFalse();
-                        form.GetTestAccessor().RevisionGrid.GetTestAccessor().VisibleRevisionCount.Should().Be(3);
+                        //// TBD timing issue form.GetTestAccessor().RevisionGrid.GetTestAccessor().VisibleRevisionCount.Should().Be(3);
```
Comment (mstv) [inline IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormBrowseTests.cs:158]: Perhaps need to wait for the reloading to have begun before `WaitForRevisionsToBeLoaded`, which checks for `IsDataLoadComplete == true`?

Comment (gerhardol) [reply IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormBrowseTests.cs:158]: I am working on the tests, they are broken and disabled in AppVeyor but need this PR first.
The GUI dropdowns are currently just init from GUI and is loaded with FileAndForget so the test is failing

**Review by RussKie — COMMENTED** [review]
Thank you for adding the comments 👍

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 6]
```diff
@@ -980,7 +980,7 @@ public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> get
                     string pathFilter = BuildPathFilter(_filterInfo.PathFilter);
                     ArgumentBuilder args = RevisionReader.BuildArguments(_filterInfo.CommitsLimit,
                         _filterInfo.RefFilterOptions,
-                        _filterInfo.BranchFilter,
+                        _filterInfo.IsShowFilteredBranchesChecked ? _filterInfo.BranchFilter : string.Empty,
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:983]: Why does the property `ByBranchFilter` not suffice? It makes the property `BranchFilter` return the empty string.

File: IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormBrowseTests.cs (hunk 5 above)
Comment (mstv) [reply IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormBrowseTests.cs:158]: `JoinPendingOperations()` in addition?

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs (hunk 6 above)
Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:983]: ByBranchFilter  is set also when IsShowCurrentBranchChecked  is set, but the contents in BranchFilter must then be ignored
I plan to change this in a followup.
First planned to do this in this PR, but I want to fix/disable tests first

File: IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormBrowseTests.cs (hunk 5 above)
Comment (gerhardol) [reply IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormBrowseTests.cs:158]: Fixed in #10168, these comments will not be merged
There were more than just awaiting FileAndForget()

**Comment by gerhardol** [comment]
#10168 should be merged before this PR ((maybe #10169 to decrease failing tests)
I would appreciate if this could be merged soon, as it blocks #10130 that blocks #10138
This is a minor regression in 4.0, that we need to ship at some point

Also, while I remember this, I would like to cleanup the filterhandling some more:
* Separate simplify-by-decoration, no-merges, boundry as well as sort options from first-parent (they can be applied there too)
* First-parents can be used together with other types of branch filters (maybe not relevant with reflog, for all?). Currently it is dominating reflog and others 
* Set order for the various branch-filters from having first-parent as dominant to reflog->ShowCurrent->FilteredBranches->All branches
* Separate ByBranchFilter from being shared CurrentFilter and FilteredBranches to only be used for BranchFilter

More changes in tests than code

**Comment by gerhardol** [comment]
Rebased and squashed

**Review by mstv — APPROVED** [review]
